### ISSUE TITLE
✨ More idiomatic Rust examples

### ISSUE THREAD

**Issue by roboteng** [dagger/dagger#5697]
### What are you trying to do?

This isn't quite a feature request, but a request for better examples.

I'm using the Rust sdk, as I want to use dagger to run my Rust project through CI. The examples do a good job of showing how to build go and js projects, but I still found it difficult to get a useable pipeline, written in Rust, for a Rust project.

I mad e Rust project, seperated into workspaces, one of which is `ci`. This is the `main.rs` file that I used to get Dagger to run with a reasonable cache, but it took a lot of effort and research. I think it would be good if we could provide this (or more realistically, something like this) as a good starting point in the examples folder.

```rust
use dagger_sdk::HostDirectoryOpts;

const PROJECT: &str = "/project";
const PROJECT_TARGET: &str = "/project/target";

#[tokio::main]
async fn main() -> eyre::Result<()> {
    let client = dagger_sdk::connect().await?;

    let host_source_dir = client.host().directory_opts(
        ".",
        HostDirectoryOpts {
            exclude: Some(vec!["target"]),
            include: None,
        },
    );

    let build_cache = client.cache_volume("target").id().await?;
    let cargo_registry = client.cache_volume("cargo").id().await?;

    let container = client
        .container()
        .from("rust:1.71.1")
        .with_mounted_directory(PROJECT, host_source_dir.id().await?)
        .with_mounted_cache(PROJECT_TARGET, build_cache);

    let cargo_home = container.env_variable("CARGO_HOME").await?;

    let container = container.with_mounted_cache(&format!("{cargo_home}/registry"), cargo_registry);

    let container = container
        .with_workdir(PROJECT)
        .with_exec(vec!["cargo", "check"]);

    let container = container
        .with_workdir(PROJECT)
        .with_exec(vec!["cargo", "test"]);

    let container =
        container
            .with_workdir(PROJECT)
            .with_exec(vec!["rustup", "component", "add", "clippy"]);

    let container = container
        .with_workdir(PROJECT)
        .with_exec(vec!["cargo", "clippy"]);

    container
        .with_workdir(PROJECT)
        .with_exec(vec!["cargo", "build", "-p", "sample"])
        .stdout()
        .await?;

    Ok(())
}
```

I'm aware that it seems like all the sdk seem to use the same examples. Having a Rust specific example in the Rust sdk would break that pattern, but I think it would be helpful to have, even if not in the examples.

### Why is this important to you?

Having this would have saved me a lot of toil

**Comment by kjuulh** [comment]
@roboteng Hi, I could definitely add some examples for building actual rust projects =D. 

Rust has some gotchas for running in a docker build process, but I could definitely add something

**Comment by kjuulh** [comment]
For now I am thinking about adding examples for a few use cases.

- CLI
- Axum
- Leptos
- Optimized build cache (I've pretty much got this nailed, rn)
- Workspaces
- build.rs
- Sccache

Anything else that would be of interest?

**Comment by roboteng** [comment]
You might run into this, but I noticed some weirdness with caching involving clippy or other components. Since rust creates the `.cargo/bin` folder, and that's also something worth caching. I'm not at my computer right now, do I can't quite check the details, though. Those are some great starting points though!

**Comment by kjuulh** [comment]
Sounds good. I may add something like the rust-cache does for github actions as well =D

**Comment by weilbith** [comment]
I would be also interested in how to run things in parallel. Like test execution and Clippy linting or formatting or a security scan. Is this just regular asynchronous stuff using the Tokio scheduler?

**Comment by kjuulh** [comment]
@weilbith Yes. you can just run them in separate tokio tasks or whatever you prefer for parallel processing and you can execute stuff in parallel
